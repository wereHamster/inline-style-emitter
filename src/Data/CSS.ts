export { CSSStyle, CSSElementStyle, CSSStyleDeclarationEx } from "./CSS/Internal/Types";
export { cssStyleRules, cssElementStyle, styleRules } from "./CSS/Parser";
export { Emitter, MemoryEmitter, DocumentEmitter } from "./CSS/Emitter";