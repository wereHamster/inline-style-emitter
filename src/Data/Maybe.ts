// -----------------------------------------------------------------------------
// For the purposes of this library Maybe<T> is defined as 'undefined | T'.
// Here be functions which deal with such types.

export type Maybe<T> = undefined | T

export const fromMaybe = <A>(a: A, m: Maybe<A>): A =>
    m === undefined ? a : m;

// Strict in the first argument!
export const maybe = <A, B>(b: B, f: (a: A) => B, m: Maybe<A>): B =>
    m === undefined ? b : f(m);

export const catMaybes = <A>(xs: Maybe<A>[]): A[] =>
    xs.reduce((acc, x) => maybe(acc, a => [...acc, a], x), []);

