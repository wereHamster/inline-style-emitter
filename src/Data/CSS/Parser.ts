import { fromMaybe, maybe } from "../Maybe";

import { SDVFontFamily, CSSFontFaceDeclarationEx } from "./Internal/SDV/SDVFontFamily";

import { CSSStyle, CSSDeclarations, CSSRuleEx, styleHash, keyframesHash,
    CSSStyleDeclarationEx, CSSElementStyle, SDVAnimationName,
    CSSKeyframe } from "./Internal/Types";



export const cssStyleRules = (style: CSSStyle): CSSRuleEx[] =>
    style instanceof CSSElementStyle ? style.cssRules : styleRules(style);


// Parse a 'CSSStyleDeclarationEx' object into a list of 'CSSRuleEx' objects.
export const cssElementStyle = (x: CSSStyleDeclarationEx): CSSElementStyle =>
    new CSSElementStyle(styleRules(x));

export const styleRules = (x: CSSStyleDeclarationEx): CSSRuleEx[] => {
    const rules = [];
    extractStyleRules(rules, x, [], []);
    return rules;
};


// -----------------------------------------------------------------------------
// extractStyleRules
//
// Extract style rules from the style object and insert them into the rules"
// list. This function is recursive.

function extractStyleRules
( rules: CSSRuleEx[]
, s: CSSStyleDeclarationEx
, conditions: string[]
, suffixes: string[]
): void {
    const cssDeclarations: CSSDeclarations = {};

    for (const k in s) {
        const v = s[k];

        if (k[0] === ":") {
            // Pseudo classes and pseudo elements.
            extractStyleRules(rules, v, conditions, concat(suffixes, [k]));
        } else if (k[0] === "@") {
            if (k.startsWith("@media")) {
                extractStyleRules(rules, v, concat(conditions, [k]), suffixes);
            } else {
                // Ignore other @-rules.
                console.warn(`extractStyleRules: Ignoring unknown @-rule "${k}"`);
            }
        } else {
            cssDeclarations[k] = processValue(rules, k, v);
        }
    }

    tellCSSStyleRule(rules, cssDeclarations, conditions, suffixes);
}


type ValueProcessor = (rules: CSSRuleEx[], v: any) => undefined | number | string | string[];
const valueProcessors = {
    fontFamily(rules: CSSRuleEx[], v: SDVFontFamily): string | string[] {
        if (typeof v === "string") {
            return v;
        } else if (Array.isArray(v)) {
            return v.map(x => typeof x === "string" ? x : tellCSSFontFaceRule(rules, x));
        } else {
            return fromMaybe("", tellCSSFontFaceRule(rules, v));
        }
    },
    animationName(rules: CSSRuleEx[], v: SDVAnimationName): string {
        if (typeof v === "string") {
            return v;
        } else {
            return tellCSSKeyframesRule(rules, v);
        }
    }
};

// XXX: Why is it allowed to return undefined? This will essentially remove
// the property from the decls. Do we want to allow that?
export const processValue = (rules: CSSRuleEx[], k: string, v: any): undefined | string | string[] =>
    fromMaybe(v, maybe(stringifyValue(k, v), f => f(rules, v), valueProcessors[k]));

const stringifyValue = (key: string, prop: number | string | string[]): number | string | string[] =>
  (typeof prop !== "number" || prop === 0 || unitlessNumbers.has(key))
    ? prop : `${prop}px`;



const concat = <T>(xs: T[], ys: T[]): T[] =>
    [].concat.call([], xs, ys);

const concatMap = <A, B>(f: (x: A) => B[], xs: A[]): B[] =>
    xs.reduce((acc, x) => concat(acc, f(x)), []);


// In some parts of the code we treat CSSRuleEx[] as a mutable reference into
// which you can write. Sortof like a writer monad. Use these functions to
// append rules to the list.

const tellCSSStyleRule = (rules: CSSRuleEx[], cssDeclarations: CSSDeclarations, conditions: string[], suffixes: string[]): void => {
    if (Object.keys(cssDeclarations).length === 0) {
        return;
    }

    rules.push({
        type: 1,
        get hash() {
            const value = styleHash(cssDeclarations, concat(conditions, suffixes));
            Object.defineProperty(this, "hash", { value });
            return value;
        },
        conditions,
        suffixes,
        cssDeclarations,
        get className() {
            const value = "s" + this.hash;
            Object.defineProperty(this, "className", { value });
            return value;
        },
    });
};

export const tellCSSFontFaceRule = (rules: CSSRuleEx[], s: CSSFontFaceDeclarationEx): string => {
    // ASSERT notEmpty(s)

    const hash = styleHash(s, []);
    const {fontFamily, cssDeclarations} = maybe(
        {fontFamily: `f-${hash}`, cssDeclarations: Object.assign({}, s, {fontFamily: `f-${hash}`})},
        fontFamily => ({fontFamily, cssDeclarations: s}),
        s.fontFamily);

    rules.push({ type: 5, hash, cssDeclarations });

    return fontFamily;
};

export const tellCSSKeyframesRule = (rules: CSSRuleEx[], keyframes: CSSKeyframe[]): string => {
    // ASSERT notEmpty(s)

    const hash = keyframesHash(keyframes);
    const animationName = `a-${hash}`;

    rules.push({ type: 7, hash, keyframes, animationName });

    return animationName;
};


// Copied from Aphrodite, which apparently copied it from React. Who did React copy it from?
const unitlessNumbers: Set<string> = (() => {
    const prefixes = ["Webkit", "ms", "Moz", "O"];
    const explode = (key: string): string[] =>
        [key, ...prefixes.map(x => x + key.charAt(0).toUpperCase() + key.substring(1))];

    return new Set(concatMap(explode, [
        "animationIterationCount",
        "borderImageOutset",
        "borderImageSlice",
        "borderImageWidth",
        "boxFlex",
        "boxFlexGroup",
        "boxOrdinalGroup",
        "columnCount",
        "flex",
        "flexGrow",
        "flexPositive",
        "flexShrink",
        "flexNegative",
        "flexOrder",
        "gridRow",
        "gridColumn",
        "fontWeight",
        "lineClamp",
        "lineHeight",
        "opacity",
        "order",
        "orphans",
        "tabSize",
        "widows",
        "zIndex",
        "zoom",

        // SVG-related properties
        "fillOpacity",
        "floodOpacity",
        "stopOpacity",
        "strokeDasharray",
        "strokeDashoffset",
        "strokeMiterlimit",
        "strokeOpacity",
        "strokeWidth",
    ]));
})();
