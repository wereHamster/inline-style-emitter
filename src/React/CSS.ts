// We're not actually importing any bindings from React, just types.
import { Props, ReactType, ReactElement, ReactNode,
    ComponentClass, StatelessComponent } from "react";

import { Maybe, maybe, catMaybes } from "../Data/Maybe";

import { CSSStyle, cssStyleRules, Emitter } from "../Data/CSS";



// -----------------------------------------------------------------------------

export class Handle {

    typeWrappers: WeakMap<any, any> = new WeakMap();
    // ^ Cache for React type wrappers. This is needed because React uses
    // reference equality to check if the type has changed or not.
    //
    // See 'wrapType' and 'mkTypeWrapper'.


    constructor
    ( private React // typeof React 
    , private emitter: Emitter
    ) {}


    private newProps = (style: CSSStyle, className: Maybe<string>): Maybe<{ style: undefined, className: string }> => {
        this.emitter.emitStyle(style);

        return {
            style: undefined,
            className: maybe("", x => `${x} `, className) + catMaybes(cssStyleRules(style).map(x =>
                x.type === 1 ? x.className : undefined)).join(" "),
        };
    }

    wrapElement = (el: null | ReactElement<any>): null | ReactElement<any> => {
        if (el === null) {
            return null;
        }

        const {React} = this;
        const type = el.type;

        if (typeof type === "function") {
            // ComponentClass<any> | StatelessComponent<any>
            return React.createElement(this.wrapType(type), el.props);

        } else if (typeof type === "string") {
            const oldProps    = <{ children?, style?, className? }> el.props
                , oldChildren = oldProps.children
                , style       = oldProps.style;

            const newChildren = oldChildren
                ? this.wrapNode(oldChildren)
                : undefined;

            // fmap over Maybe... 
            const newProps = style !== undefined
                ? this.newProps(style, oldProps.className)
                : undefined;

            // Avoid cloning the element if neither the props nor the children have
            // changed.
            return newProps === undefined && newChildren === undefined
                ? el : React.cloneElement(el, newProps, newChildren);

        } else {
            // Ohoh, the type is neither a function nor a string?
            throw new TypeError(`wrapElement: unknown type ${typeof type}`);
        }
    };

    private wrapNode = (node: ReactNode): ReactNode => {
        if (Array.isArray(node)) {
            return this.React.Children.map(node, this.wrapNode);
        } else if (this.React.isValidElement(node)) {
            return this.wrapElement(<any>node);
        } else { // string | number | boolean | {}
            return node;
        }
    };

    // Wrap a component class or stateless component in a new type so that we can
    // process its children.
    private wrapType = (type: ComponentClass<any> | StatelessComponent<any>): ReactType => {
        let w = this.typeWrappers.get(type);

        if (w === undefined) {
            w = this.mkTypeWrapper(type);
            this.typeWrappers.set(type, w);
        }

        return w;
    };

    private mkTypeWrapper = (type: ComponentClass<any> | StatelessComponent<any>): ReactType => {
        const {React, wrapElement} = this;

        if (type.prototype && type.prototype instanceof React.Component) {
            return <any> class extends (<ComponentClass<any>>type.prototype.constructor) {
                render() { return wrapElement(super.render()); }
            };
        } else {
            return (props, context) => wrapElement((<any>type)(props, context));
        }
    };
}


export const newHandle = (React: any, emitter: Emitter): Handle =>
    new Handle(React, emitter);

export const wrapElement = (h: Handle, el: null | ReactElement<any>): null | ReactElement<any> =>
    h.wrapElement(el);