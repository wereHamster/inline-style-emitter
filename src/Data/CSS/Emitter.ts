import { Maybe, catMaybes } from "../Maybe";

import { CSSRuleExHash, CSSRuleEx, cssRuleExText, CSSStyleDeclarationEx,
    CSSElementStyle, CSSStyle } from "./Internal/Types";
import { cssStyleRules } from "./Parser";



// -----------------------------------------------------------------------------
// The Emitter is an interface which implements the logic of actually writing
// a style object into a target location. This can be either directly into a DOM
// CSSStyleSheet objects or into a file in the case of server-side rendering.

export interface Emitter {
    emitRule(rule: CSSRuleEx): void;
    emitStyle(style: CSSStyle): void;
}



// -----------------------------------------------------------------------------
// The 'MemoryEmitter' stores the rules in memory. You can at any point access
// the list of emitted CSS rules.
//
// It is useful for server-side rendering.  

export class MemoryEmitter implements Emitter {

    cssRules: Map<CSSRuleExHash, CSSRuleEx> = new Map();
    // ^ All CSS rules which have been emitted using this emitter. Treat
    // as read-only when you access it directly. It can only be modified
    // by the 'emitStyle' function.

    emitRule = (rule: CSSRuleEx): void => {
        this.cssRules.set(rule.hash, rule);
    }

    emitStyle = (style: CSSStyle): void => {
        cssStyleRules(style).map(this.emitRule);
    }
}

export const newMemoryEmitter = (): MemoryEmitter =>
    new MemoryEmitter();



// -----------------------------------------------------------------------------
// An DocumentEmitter writes the styles directly into the DOM. It creates
// a new CSSStyleSheet object and inserts the rules into that.
//
// Rules are only ever inserted, old ones are not removed. This is a problem
// for <IE9 which has a limit on how many CSS rules it can have. And it wastes
// a bit of memory because we keep rules around which are potentially not
// needed.
//
// However, if the application doesn't use dynamic styles (eg. theming), then
// there is a fixed upper bound which the application will slowly approach but
// never exceed.
//
// It might still be useful to remove unused rules, however we should first
// measure the performance of the two approaches and then decide whether the
// tradeoff is worth it or not.

const supportedProperties = (() => {
    const rmsPrefix = /^-ms-/;
    const rdashAlpha = /-([a-z])/g;
    const fcamelCase = (_, l) => l.toUpperCase();
    const camelCase = (x: string): string =>
        x.replace(rmsPrefix, "ms-").replace(rdashAlpha, fcamelCase);

    return (window: Window): Set<string> => {
        const computedStyle = window.getComputedStyle(document.documentElement, "");
        return new Set(Array.from(computedStyle).map(camelCase));
    };
})();

export class DocumentEmitter implements Emitter {

    private styleSheet: CSSStyleSheet;
    // ^ The DOM StyleSheet into which we insert the rules. It is created
    // in the constructor.

    private cssRules: Set<string> = new Set();
    // ^ We keep track of which rules (based on their hash) we've already
    // inserted. This is needed to avoid inserting the same rule multiple times.
    //
    // The CSSStyleSheet doesn't provide a good interface for determining if
    // a particular rule is present or not.


    private supportedProperties: Set<string> = new Set();
    // ^ The set of CSS properties which are supported by the document.

    propertyIsSupported = (prop: string): boolean =>
        this.supportedProperties.has(prop);


    constructor(private window: Window) {
        const {document} = window;

        const style = document.createElement("style");
        style.type = "text/css";
        document.head.appendChild(style);

        this.styleSheet = <CSSStyleSheet> document.styleSheets[document.styleSheets.length - 1];

        this.supportedProperties = supportedProperties(window);
    }


    emitRule = (rule: CSSRuleEx): void => {
        const {hash} = rule;

        if (!this.cssRules.has(hash)) {
            this.cssRules.add(hash);
            this.styleSheet.insertRule(cssRuleExText(this, rule),
                this.styleSheet.cssRules.length);
        }
    }

    emitStyle = (style: CSSStyle): void => {
        cssStyleRules(style).map(this.emitRule);
    }
}

export const newDocumentEmitter = (window: Window): DocumentEmitter =>
    new DocumentEmitter(window);