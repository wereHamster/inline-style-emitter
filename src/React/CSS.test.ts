/* global describe, it, expect */

declare var describe: any, it: any, expect: any;

import * as React from "react";
import { createElement } from "react";
import renderer from "react-test-renderer";

import { cssElementStyle } from "../Data/CSS";
import { MemoryEmitter } from "../Data/CSS/Emitter";
import { newHandle, wrapElement } from "./CSS";




function testSnapshots1(inputs, f) {
    describe("snapshots", () =>
        inputs.forEach(input =>
            it(JSON.stringify(input), () =>
                expect(f(input)).toMatchSnapshot())));
}



describe("createElement + cssElementStyle", () => {
    testSnapshots1(
        [ {}
        , {color: "red"}
        , {color: "red", ":hover": {}}
        , {color: "red", ":hover": {color: "blue"}}
        , {"@media (max-width:1px)": {}}
        , {"@media (max-width:1px)": {color: "red"}}
        ], obj => renderer.create(createElement("div", {style: cssElementStyle(obj)})).toJSON());
});

describe("wrapElement", () => {
    const seH = newHandle(React, new MemoryEmitter());

    const div = key => createElement("div", { key });
    const fragment = [div(1), div(2), div(3)];

    const nullType = () => null;
    const componentType = ({children}: {children?: any}) =>
        createElement("div", {children});


    // Test all possible combinations of type and children.
    describe("snapshots", () => {
        [nullType, componentType].forEach((type, typeI) => {
            [[], [null], [false], fragment, [fragment]].forEach((children, childrenI) => {
                it(JSON.stringify({typeI, childrenI}), () => {
                    const el = wrapElement(seH, createElement(type, {}, ...children));
                    expect(renderer.create(el).toJSON()).toMatchSnapshot();
                });
            });
        });
    });


    it("should propagate null", () => {
        expect(wrapElement(seH, null)).toBe(null);
    });
    it("should not fail when the element render function returns null", () => {
        expect(() => wrapElement(seH, createElement(nullType)))
            .not.toThrow();
    });
    it("should not fail when the children are passed as spread arguments", () => {
        expect(() => wrapElement(seH, createElement(nullType, {}, ...fragment)))
            .not.toThrow();
    });
    it("should not fail when the children are an array", () => {
        expect(() => wrapElement(seH, createElement(nullType, {}, fragment)))
            .not.toThrow();
    });
    it("should append to className if it is already set", () => {
        const div = createElement("div", {
            style: cssElementStyle({color: "red"}),
            className: "icon-x",
        });

        expect(renderer.create(wrapElement(seH, div)).toJSON().props.className)
            .toBe("icon-x s29ec1a4d45b6d719");
    });
});
