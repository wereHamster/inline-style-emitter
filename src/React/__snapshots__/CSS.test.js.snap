exports[`createElement + cssElementStyle snapshots {"@media (max-width:1px)":{"color":"red"}} 1`] = `
<div
  style={
    CSSElementStyle {
      "cssRules": Array [
        Object {
          "className": "s43660e9eeeba4977",
          "conditions": Array [
            "@media (max-width:1px)"
          ],
          "cssDeclarations": Object {
            "color": "red"
          },
          "hash": "43660e9eeeba4977",
          "suffixes": Array [],
          "type": 1
        }
      ]
    }
  } />
`;

exports[`createElement + cssElementStyle snapshots {"@media (max-width:1px)":{}} 1`] = `
<div
  style={
    CSSElementStyle {
      "cssRules": Array []
    }
  } />
`;

exports[`createElement + cssElementStyle snapshots {"color":"red",":hover":{"color":"blue"}} 1`] = `
<div
  style={
    CSSElementStyle {
      "cssRules": Array [
        Object {
          "className": "s46576037294a675a",
          "conditions": Array [],
          "cssDeclarations": Object {
            "color": "blue"
          },
          "hash": "46576037294a675a",
          "suffixes": Array [
            ":hover"
          ],
          "type": 1
        },
        Object {
          "className": "s29ec1a4d45b6d719",
          "conditions": Array [],
          "cssDeclarations": Object {
            "color": "red"
          },
          "hash": "29ec1a4d45b6d719",
          "suffixes": Array [],
          "type": 1
        }
      ]
    }
  } />
`;

exports[`createElement + cssElementStyle snapshots {"color":"red",":hover":{}} 1`] = `
<div
  style={
    CSSElementStyle {
      "cssRules": Array [
        Object {
          "className": "s29ec1a4d45b6d719",
          "conditions": Array [],
          "cssDeclarations": Object {
            "color": "red"
          },
          "hash": "29ec1a4d45b6d719",
          "suffixes": Array [],
          "type": 1
        }
      ]
    }
  } />
`;

exports[`createElement + cssElementStyle snapshots {"color":"red"} 1`] = `
<div
  style={
    CSSElementStyle {
      "cssRules": Array [
        Object {
          "className": "s29ec1a4d45b6d719",
          "conditions": Array [],
          "cssDeclarations": Object {
            "color": "red"
          },
          "hash": "29ec1a4d45b6d719",
          "suffixes": Array [],
          "type": 1
        }
      ]
    }
  } />
`;

exports[`createElement + cssElementStyle snapshots {} 1`] = `
<div
  style={
    CSSElementStyle {
      "cssRules": Array []
    }
  } />
`;

exports[`wrapElement snapshots {"typeI":0,"childrenI":0} 1`] = `null`;

exports[`wrapElement snapshots {"typeI":0,"childrenI":1} 1`] = `null`;

exports[`wrapElement snapshots {"typeI":0,"childrenI":2} 1`] = `null`;

exports[`wrapElement snapshots {"typeI":0,"childrenI":3} 1`] = `null`;

exports[`wrapElement snapshots {"typeI":0,"childrenI":4} 1`] = `null`;

exports[`wrapElement snapshots {"typeI":1,"childrenI":0} 1`] = `<div />`;

exports[`wrapElement snapshots {"typeI":1,"childrenI":1} 1`] = `<div />`;

exports[`wrapElement snapshots {"typeI":1,"childrenI":2} 1`] = `<div />`;

exports[`wrapElement snapshots {"typeI":1,"childrenI":3} 1`] = `
<div>
  <div />
  <div />
  <div />
</div>
`;

exports[`wrapElement snapshots {"typeI":1,"childrenI":4} 1`] = `
<div>
  <div />
  <div />
  <div />
</div>
`;
